### ISSUE TITLE
Inclusion request apache-apr/1.6.3@jgsogo/stable

### ISSUE THREAD

**Issue by lasote** [jgsogo/conan-apache-apr#2]
## Inclusion Request Review

This is a review for [apache-apr/1.6.3@jgsogo/stable (jgsogo/conan-packages)](https://github.com/jgsogo/conan-apache-apr) inclusion request to conan-center.
This review follows the [inclusion guidelines or third party libraries](http://docs.conan.io/en/latest/uploading_packages/bintray/conan_center_guide.html#inclusion-guidelines-for-third-party-libraries).

### Package
apache-apr/1.6.3@jgsogo/stable

#### Recipe quality
- [x] Git public repository
- [ ] Recipe fields
Description missing.
- [x] Linter
- [x] Updated
- [x] Clean
- [ ] test_package

Not working because of missing packages for OSX

- [x] LICENSE of the recipe
- [ ] LICENSE of the library

Missing package the license of the library in every bin package

#### CI Integration
- [ ] (If not header only) CI configurations for Linux/MacOSX/Windows

No binary packages for many platforms. No windows, No debug packages?

#### Bintray Package Information
- [x] Description
- [x] Licenses
- [ ] Tags
- [ ] Maturity
- [ ] Website (if any)
- [ ] Issue tracker
- [x] Version control

### Additional comments

There are some outdated packages, clean them up with: 
`conan remove apache-apr/1.6.3@jgsogo/stable --outdated -r bincrafters`

It looks like a C library, but you are not removing the `compiler.libcxx` setting:
```
      def configure(self):
          del self.settings.compiler.libcxx
```

Packages matrix:
![image](https://user-images.githubusercontent.com/3807515/37196954-10638210-2379-11e8-9792-2022e0684c94.png)

**Comment by jgsogo** [comment]
Much better now. There were a couple of errors (one of them in `conan_sword_and_sorcery`) but now I think it has much better looking:

 * [x] Recipe fields: description
 * [x] Macos: apple-clang 7.3 <-- I don't know what you mean in your comment.
 * [x] LICENSE of the library inside the package
 * [x] CI for win/linux/mac (bug fixed in `conan_sword_and_sorcery`)
 * [x] Bintray info: maturity
 * [x] Bintray info: website
 * [x] Removed outdated packages
 * [x] Removed `libcxx` from settings
 * [x] Updated matrix:

![imagen](https://user-images.githubusercontent.com/1406456/37223962-6c22efe2-23d2-11e8-894c-47a540b081aa.png)

Questions:
 * any suggestion for tags?
 * issue tracker, should I point to [Apache's Jira](https://issues.apache.org/jira/secure/Dashboard.jspa)?

---

About the matrix: windows seems to encode headers as `<build_type> <arch>` while unix makes them as `<arch> <build_type>`, is it ok?

**Comment by jgsogo** [comment]
Wait! Just realized a bug: clang jobs are being compiled with gcc (https://travis-ci.org/jgsogo/conan-apache-apr/jobs/351413547#L742). :/

**Comment by jgsogo** [comment]
Fixed! [Here it is](https://travis-ci.org/jgsogo/conan-apache-apr/jobs/351413547#L333)... but maybe others building with AutoTools have the same issue... maybe it is better to remove `gcc` compiler from those dockers.

**Comment by jgsogo** [comment]
I managed to fix it defining CC and CXX environment variables in the profile files I'm using to run `conan create` (it is a fix/workaround in my `conan_sowrd_and_sorcey` that fits your docker images... vamos, una ñapa).

Go ahead!

**Comment by jgsogo** [comment]
@lasote... I think it deserves a new iteration.

**Comment by lasote** [comment]
haha, yes, It deserves it. On my way!

**Comment by lasote** [comment]
- Missing binaries for apple-clang > 7.3 (8.1 and 9.0)

![image](https://user-images.githubusercontent.com/3807515/37411483-304c1f50-27a3-11e8-9c98-23aa0bf58f21.png)

- Why move all the installed files to subfolder `apr-1`? Will it require to change the way people include the headers?


- issue_tracker_url field missing in bintray package page:

![image](https://user-images.githubusercontent.com/3807515/37411600-724f1d94-27a3-11e8-8d0f-a6e4038d73e9.png)

**Comment by jgsogo** [comment]
* [ ] Running right now with more apple-clang jobs
 * [ ] Linux already copies includes into a `apr-1` folder
 * [x] About the issue tracker, I asked you before: should I point to [Apache's Jira](https://issues.apache.org/jira/secure/Dashboard.jspa)?

**Comment by lasote** [comment]
The issue tracker should be `https://github.com/jgsogo/conan-apache-apr/issues`.
About the `apr-1` folder, I don't get it. So the Linux builds copy the things to `apr-1` but in other OS is in the root? WTF? My concern is, if you search for examples of this library (http://dev.ariel-networks.com/apr/apr-tutorial/sample/mp-sample.c) the includes point directly to the root:

```
#include <apr_general.h>
#include <apr_pools.h>
```
So the examples won't compile with your package.

**Comment by jgsogo** [comment]
As you can see, the linux stuff is not modified at all during `build` stage... in fact Windows with CMake seems to be _experimental_ ([look here](https://github.com/apache/apr/blob/trunk/README.cmake)).

Options: 
 * Use VS to build [link](https://github.com/apache/apr/blob/trunk/README#L155)
 * keep Windows unmodified and add an extra include_dir for Linux or Windows
 * keep these modifications for Windows and do not add any additional include_dir.

Although, if examples uses them without the `apr-1`, the final solution must go in that direction... we should look for up-to-date examples.

**Comment by lasote** [comment]
We could do a trick: `self.cpp_info.includedirs.append("apr-1")` so the includes will be located without the prefix too 📦

**Comment by jgsogo** [comment]
Yes, and the `test_package` example, what do we put in the example? XD

Ok, I can do it and remove customizations, but there is a pending task: look for examples, tutorials,... and wait for version 2.0 to _fix_ it.

**Comment by lasote** [comment]
> Yes, and the test_package example, what do we put in the example? XD

Mix them if you want ;) one with "apr-1/" and other without it. It really doesn't matter that way.

**Comment by jgsogo** [comment]
I have to give it a try, because it has to be compatible with the [apache-apr-util](https://github.com/jgsogo/conan-apache-apr-util) recipe and with the [apache-log4cxx](https://github.com/jgsogo/conan-apache-log4cxx) one :/

**Comment by lasote** [comment]
Sure, check it, but we are just -adding- an additional include directory `self.cpp_info.includedirs.append("apr-1")`, the default one is to the `package/include` and we add other to the `package/include/apr-1`

**Comment by jgsogo** [comment]
I'm thinking about removing all the customization for Windows, but I don't know if it worths it...

**Comment by jgsogo** [comment]
Will go without the `apr-1` extra directory as it is now so consumer will use `#include "apr.h"`, so I have to modify recipe for linux:
 * for CMake add the nested folder into includedirs (may remove the default `<package>/include` as it is not needed)
 * for pkg-config work following this approach https://github.com/conan-io/conan/issues/2619#issuecomment-373744305 to get rid of abs path for include prefix.

**Comment by jgsogo** [comment]
Done! Although I'm facing a weird error for apple-clang 9.0 (7.3 and 8.1 works fine). Is there something wrong with `osx_image: xcode9.0`?

**Comment by lasote** [comment]
Use  `xcode9.2`

**Comment by jgsogo** [comment]
@lasote, let's see if it all is ok and we can [move forwards](https://github.com/jgsogo/conan-apache-apr-util).

**Comment by lasote** [comment]
I've an issue on OSX:

```
apache-apr/1.6.3@jgsogo/stable (test package) imports(): Copied 2 '.dylib' files: libapr-1.0.dylib, libapr-1.dylib
apache-apr/1.6.3@jgsogo/stable (test package): Running build()
-- The CXX compiler identification is AppleClang 9.0.0.9000039
-- Check for working CXX compiler: /usr/local/opt/ccache/libexec/c++
-- Check for working CXX compiler: /usr/local/opt/ccache/libexec/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Conan: called by CMake conan helper
-- Conan: Using cmake global configuration
-- Conan: Adjusting default RPATHs Conan policies
-- Conan: Adjusting language standard
-- Conan: C++ stdlib: libc++
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/luism/workspace/inclusions/storage/jgsogo/conan-packages/apache-apr__1.6.3__jgsogo__stable/git_clone/pkg/test_package/build/7dd9e1a83ffe63c45af7c671c6032d748dc83894
Scanning dependencies of target example
[ 50%] Building CXX object CMakeFiles/example.dir/example.cpp.o
[100%] Linking CXX executable bin/example
[100%] Built target example
apache-apr/1.6.3@jgsogo/stable (test package): Running test()
dyld: Library not loaded: /Users/travis/.conan/data/apache-apr/1.6.3/jgsogo/stable/package/389adbab17f57e67170eb9b6c367e3676bd95d7c/lib/libapr-1.0.dylib
Referenced from: /Users/luism/workspace/inclusions/storage/jgsogo/conan-packages/apache-apr__1.6.3__jgsogo__stable/git_clone/pkg/test_package/build/7dd9e1a83ffe63c45af7c671c6032d748dc83894/bin/./example
Reason: Incompatible library version: example requires version 7.0.0 or later, but libapr-1.0.dylib provides version 5.0.0
ERROR: PROJECT: Error in test() method, line 23
self.run(".%sexample" % os.sep)
ConanException: Error 6 while executing ./example
```

I'm not fully sure if it is related, but I noticed that you are packaging static and shared libraries, but the test package is always importing to the bin the dylibs, so, how do you know which library is being linked? I would suggest the usage of `build_id` instead of `package_id` to build once but package twice.  I think it's safer to package only the shared or only the static.

**Comment by jgsogo** [comment]
I know very little about OSX, but I have activated shared/static build in Travis jobs and [they are succeeding](https://travis-ci.org/jgsogo/conan-apache-apr/jobs/355533964), so maybe there is something wrong in your configuration. Can you help me to reproduce your error?

Anyway, I will have a look at `build_id` and try to use it.

**Comment by lasote** [comment]
The errors with OSX are gone, I don't know what changed but it is working ok now, maybe a bug in conan already solved? 
So, we have still some pending issues:

- Missing packages for Visual Studio 14
- Missing packages for apple-clang 9.1 (can be added to travis with xcode 9.3)
- Missing packages for clang 4.0
- Complete the "labels" field in the bintray account.

**Comment by jgsogo** [comment]
Can you check it all again? I think that everything is ok now.

This is the matrix I've just generated using current CI files:

![image](https://user-images.githubusercontent.com/1406456/42178129-7fe5e824-7e2f-11e8-9e05-0b1f801ab156.png)

**Comment by lasote** [comment]
Why x86 only for the latest compilers and why not debug packages except for visual studio?

**Comment by jgsogo** [comment]
I copy these CI conf files from one project to another. I like to have a little of each setting and option available, but no heuristic or special reason to choose one over another.

This project doesn't last too much to compile (<40min in appveyor) so I can fill a little bit more the matrix. Anyway, it will always be gaps in it, do Conan log most used configurations in order to have those binaries available at conan-center?

**Comment by jgsogo** [comment]
@lasote, kindly reminder... 🙃

**Comment by lasote** [comment]
>  do Conan log most used configurations in order to have those binaries available at conan-center?

No, we are not using stats to choose which configuration we build. 

The typical matrix would be:

![image](https://user-images.githubusercontent.com/3807515/43067123-b4101060-8e66-11e8-8ee4-38ed700a625a.png)

**Comment by jgsogo** [comment]
Ok, I'll create all those CI jobs ^^

**Comment by jgsogo** [comment]
@lasote I've been trying minGW for this package, but I don't know if it will worth the effort, [here is a link](http://mingw.5.n7.nabble.com/Apache-Apr-td26683.html) with some patches, but those do not correspond to the current release version. 

I think I will add more CI jobs to generate more binaries, but with respect to MinGW I will leave here an issue opened and a branch. If you consider that it is needed to submit the recipe to conan-center... well, people can always add more remotes :smile: and I think apache-apr is not a demanded recipe, it is better to spent time working on other recipes.

**Comment by lasote** [comment]
Yeah, no problem. As you said, it is better to wait for feedback or other configurations request.
I'm taking some days off for summer, so sorry if I don't answer ;)
